tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go

```diff
@@ -0,0 +1,104 @@
+
+// Analyzer is the analyzer.
+var Analyzer = &analysis.Analyzer{
+	Name:     "declarationcommentmismatch",
+	Doc:      doc,
+	Requires: []*analysis.Analyzer{inspect.Analyzer},
+	Run:      run,
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	inspectResult := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	nodeFilter := []ast.Node{
+		(*ast.File)(nil),
+		(*ast.FuncDecl)(nil),
+		(*ast.ValueSpec)(nil),
+		(*ast.TypeSpec)(nil),
+		(*ast.GenDecl)(nil),
+	}
+
+	inspectResult.Nodes(nodeFilter, func(n ast.Node, push bool) bool {
+		if !push {
+			return true
+		}
+
+		if astFile, ok := n.(*ast.File); ok {
+			file := pass.Fset.File(astFile.Pos())
+			if file != nil && strings.HasSuffix(file.Name(), "_test.go") {
+				return false
+			}
+			return true
+		}
+
+		if astFunction, ok := n.(*ast.FuncDecl); ok {
+			checkCommentCaseMatches(pass, astFunction.Doc, astFunction.Name.String(), "function", astFunction.Pos())
+			return true
+		}
+
+		if astVar, ok := n.(*ast.ValueSpec); ok {
+			checkCommentCaseMatches(pass, astVar.Doc, astVar.Names[0].String(), "variable", astVar.Pos())
+			return true
+		}
+
+		if astGenDecl, ok := n.(*ast.GenDecl); ok {
+			documentation := astGenDecl.Doc
+			for _, spec := range astGenDecl.Specs {
+				if astType, ok2 := spec.(*ast.TypeSpec); ok2 {
+					switch astType.Type.(type) {
+					case *ast.ArrayType:
+						checkCommentCaseMatches(pass, documentation, astType.Name.String(), "array", astType.Pos())
+					case *ast.StructType:
+						checkCommentCaseMatches(pass, documentation, astType.Name.String(), "struct", astType.Pos())
+					case *ast.InterfaceType:
+						checkCommentCaseMatches(pass, documentation, astType.Name.String(), "interface", astType.Pos())
+					case *ast.MapType:
+						checkCommentCaseMatches(pass, documentation, astType.Name.String(), "map", astType.Pos())
+					case *ast.ChanType:
+						checkCommentCaseMatches(pass, documentation, astType.Name.String(), "channel", astType.Pos())
+					}
+					return true
+				}
+			}
+		}
+
+		return true
+	})
+	return nil, nil
+}
+
+func checkCommentCaseMatches(pass *analysis.Pass, doc *ast.CommentGroup, objectName string, objectType string, position token.Pos) {
+	if doc == nil {
+		return
+	}
+	commentSplit := strings.Split(doc.List[0].Text, " ")
+	if len(commentSplit) < 2 {
+		return
+	}
+
+	firstObjectLetter := []rune(objectName[0:1])[0]
+
+	firstCommentWord := commentSplit[1]
+	firstCommentLetter := []rune(firstCommentWord[0:1])[0]
+
+	if firstCommentWord[1:] == objectName[1:] && ((unicode.IsLower(firstObjectLetter) && unicode.ToUpper(firstObjectLetter) == firstCommentLetter) || (unicode.IsLower(firstCommentLetter) && unicode.ToUpper(firstCommentLetter) == firstObjectLetter)) {
+		pass.Report(analysis.Diagnostic{
+			Pos:     position,
+			Message: fmt.Sprintf("Mismatching capitalization for %s %s and comment starting with %s", objectType, objectName, firstCommentWord),
```

Comment:
I like to quote the `%s` using single quotes to make them better visible (what if one of them is empty?). But before doing this, please check if we follow one style of doing such things and if yes, then maybe stick to it.

Reply:
Just checked other linters and there the `%s`'s were not quoted so I'll leave them as is. Plus those `%s`'s cannot be empty logibally speaking, or `pass.Report` would never be reached.

Reply:
This is fine not to quote this.

I am having a second though here. This message will be displayed to the user, so maybe it should include some more information. I am thinking of:

1. Telling the user that this has happened because the first word of the comment is the same as the function* name
2. Hinting how to fix that - we will not see any problems where the comment needs to be changed from `// foo` into `// Foo` - this will be handled by govet. So all cases that we discover are about changing `// Foo` into `// foo`. Maybe we can say it directly? Example: _This comment should start with non-capital rune (<explanation why>)_

Reply:
Good idea. I've done it for both cases; even if govet already covers half it can't hurt. I've now also changed the position in the file shown to the user from the position of the declaration to the position of the comment. I think that makes more sense now that I directly recommend what change to make.

Reply:
Now it only checks for cases where the function is not capitalized